[SPARK-48658][SQL] Encode/Decode functions report coding errors instead of mojibake for  unmappable characters

yaooqinn opened the pull request:
### What changes were proposed in this pull request?
This PR makes encode/decode functions report coding errors instead of mojibake for unmappable characters, take `select encode('渭城朝雨浥轻尘', 'US-ASCII')` as an example

Before this PR，

```sql
???????
```
After this PR，
```json
org.apache.spark.SparkRuntimeException
{
  "errorClass" : "MALFORMED_CHARACTER_CODING",
  "sqlState" : "22000",
  "messageParameters" : {
    "charset" : "US-ASCII",
    "function" : "`encode`"
  }
}
```

### Why are the changes needed?

Improve data quality. 


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes.

When set spark.sql.legacy.codingErrorAction to true, encode/decode functions replace unmappable characters with mojibake instead of reporting coding errors.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

new unit tests
### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->no

yaooqinn committed:
[SPARK-48658][SQL] Encode/Decode functions report coding error instead of mojibake

yaooqinn committed:
[SPARK-48658][SQL] Encode/Decode functions report coding error instead of mojibake

yaooqinn committed:
[SPARK-48658][SQL] Encode/Decode functions report coding error instead of mojibake

yaooqinn committed:
fix ExplainSuite

yaooqinn committed:
fix golden file tests

yaooqinn committed:
fix golden file tests

yaooqinn commented:
cc @cloud-fan @dongjoon-hyun @LuciferYang @HyukjinKwon thanks

cloud-fan commented on connector/connect/common/src/test/resources/query-tests/explain-results/function_decode.explain:
```diff
@@ -1,2 +1,2 @@
-Project [decode(cast(g#0 as binary), UTF-8, false) AS decode(g, UTF-8)#0]
+Project [decode(cast(g#0 as binary), UTF-8, false, false) AS decode(g, UTF-8)#0]
```
shall we exclude these two flags from the pretty string of the expression?

yaooqinn replied:
OK

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:
```diff
@@ -2724,35 +2729,38 @@ case class StringDecode(bin: Expression, charset: Expression, legacyCharsets: Bo
-        }
-      """
-    })
+    val expr = ctx.addReferenceObj("this", this)
```
since we are here, shall we re-implement this expression with `StaticInvoke`?

yaooqinn committed:
Encode RuntimeReplaceable with StaticInvoke

yaooqinn committed:
Decode RuntimeReplaceable with StaticInvoke

yaooqinn committed:
fix tests

yaooqinn committed:
Merge branch 'master' into SPARK-48658

yaooqinn committed:
fix

yaooqinn commented:
Hi @cloud-fan, I have addressed your comments. The expressions are now replaced at runtime by static invoke, and the string representations no longer contain those legacy flags.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala:
```diff
@@ -107,7 +107,7 @@ class CodeGenerationSuite extends SparkFunSuite with ExpressionEvalHelper {
       strExpr = StringDecode(Encode(strExpr, "utf-8"), "utf-8")
```
can we use a different expression for testing? The codegen size is greatly decreased after using `StaticInvoke` in `Encode`.

cloud-fan replied:
e.g. `StringTrim`

yaooqinn replied:
Nice catch!

yaooqinn commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -5010,6 +5010,14 @@ object SQLConf {
+      "instead of reporting coding errors.")
+    .version("4.0.0")
+    .booleanConf
+    .createWithDefault(false)
```
The reasons I'd like to make it independent of ANSI are:
- Part of the implication of ANSI is Hive-incompatibility,
- Hive also reports coding errors, so it was a mistake when we ported this from hive
- These functions are not ANSI-defined
- The error behaviors are also not found in ANSI

The reasons mentioned above indicate that this behavior is more of a legacy trait of Spark itself.

yaooqinn committed:
address comments

yaooqinn commented:
Merged to master.

Thank you @cloud-fan @HyukjinKwon for the help
